File: haskell/009/euler.hs
import Data.List

main :: IO ()

findPythagoreanTriple :: (Integral a, Show a) => a -> a -> a -> [a]
findPythagoreanTriple sum m n
    | sum  < 12 = error ("findPythagoreanTriple " ++ (show sum))
    | m < sum && n < sum - 1 && m > n && a + b + c == sum && (a * a) + (b * b) == (c * c) = [a, b, c]
    | m < sum && n < sum - 1 && m > n = findPythagoreanTriple sum m (n + 1)
    | m < sum && n < sum - 1          = findPythagoreanTriple sum (m + 1) 1
    | otherwise                       = []
    where
      a = (m * m) - (n * n)
      b = 2 * m * n
      c = (m * m) + (n * n)

pythagoreanTripleWithSum :: (Integral a, Show a) => a -> [a]
pythagoreanTripleWithSum sum = findPythagoreanTriple sum 2 1

main = print(product (pythagoreanTripleWithSum 1000))

File: erlang/009/euler.erl
#!/usr/bin/env escript

main(_) ->
  io:fwrite("~w~n", [lists:foldl(fun(X, Product) -> X * Product end, 1, pythagorean_triple_with_sum(1000))]).

pythagorean_triple_with_sum(Sum) when Sum > 11 ->
  pythagorean_triple_with_sum(Sum, 2, 1).
pythagorean_triple_with_sum(Sum, M, N) when M < Sum, N < Sum - 1, M > N ->
  A = (M * M) - (N * N),
  B = 2 * M * N,
  C = (M * M) + (N * N),
  if
    A + B + C == Sum, (A * A) + (B * B) == (C * C) ->
      [A, B, C];
    true ->
      pythagorean_triple_with_sum(Sum, M, N + 1)
  end;
pythagorean_triple_with_sum(Sum, M, N) when M < Sum, N < Sum - 1 ->
  pythagorean_triple_with_sum(Sum, M + 1, 1);
pythagorean_triple_with_sum(_, _, _) ->
  [].
